### ISSUE TITLE
Failing test - AttributeError: module 'jinja2.ext' has no attribute 'with_'

### ISSUE THREAD

**Issue by onny** [django-compressor/django-compressor#1139]
Hey,
on NixOS I'm trying to update the django-compressor package https://github.com/NixOS/nixpkgs/pull/186298 . Unfortunately many tests still fail. One of them is `OfflineCompressBasicTestCase.test_deleting_manifest_does_not_affect_rendering`:

```
=================================== FAILURES ===================================
_ OfflineCompressBasicTestCase.test_deleting_manifest_does_not_affect_rendering _

self = <compressor.tests.test_offline.OfflineCompressBasicTestCase testMethod=test_deleting_manifest_does_not_affect_rendering>

    def setUp(self):
        # Reset template dirs, because it enables us to force compress to
        # consider only a specific directory (helps us make true,
        # independent unit tests).
        # Specify both Jinja2 and Django template locations. When the wrong
        # engine is used to parse a template, the TemplateSyntaxError will
        # cause the template to be skipped over.
        # We've hardcoded TEMPLATES[0] to be Django templates backend and
        # TEMPLATES[1] to be Jinja2 templates backend in test_settings.
        TEMPLATES = copy.deepcopy(settings.TEMPLATES)
    
        django_template_dir = os.path.join(
            TEMPLATES[0]['DIRS'][0], self.templates_dir)
        jinja2_template_dir = os.path.join(
            TEMPLATES[1]['DIRS'][0], self.templates_dir)
    
        TEMPLATES[0]['DIRS'] = [django_template_dir]
        TEMPLATES[1]['DIRS'] = [jinja2_template_dir]
    
        override_settings = {
            'TEMPLATES': TEMPLATES,
            'COMPRESS_ENABLED': True,
            'COMPRESS_OFFLINE': True
        }
    
        if 'jinja2' in self.engines:
            override_settings['COMPRESS_JINJA2_GET_ENVIRONMENT'] = (
                lambda: self._get_jinja2_env())
    
        if self.additional_test_settings is not None:
            override_settings.update(self.additional_test_settings)
    
        self.override_settings = self.settings(**override_settings)
        self.override_settings.__enter__()
    
        if 'django' in self.engines:
            self.template_path = os.path.join(
                django_template_dir, self.template_name)
    
            origin = Origin(name=self.template_path,  # Absolute path
                            template_name=self.template_name)  # Loader-relative path
            with io.open(self.template_path,
                         encoding=self.CHARSET) as file_:
                self.template = Template(file_.read(), origin=origin)
    
        if 'jinja2' in self.engines:
            self.template_path_jinja2 = os.path.join(
                jinja2_template_dir, self.template_name)
>           jinja2_env = override_settings['COMPRESS_JINJA2_GET_ENVIRONMENT']()
compressor/tests/test_offline.py:148: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compressor/tests/test_offline.py:127: in <lambda>
    lambda: self._get_jinja2_env())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <compressor.tests.test_offline.OfflineCompressBasicTestCase testMethod=test_deleting_manifest_does_not_affect_rendering>

    def _get_jinja2_env(self):
        import jinja2.ext
        from compressor.offline.jinja2 import url_for, SpacelessExtension
        from compressor.contrib.jinja2ext import CompressorExtension
    
        # Extensions needed for the test cases only.
        extensions = [
            CompressorExtension,
            SpacelessExtension,
>           jinja2.ext.with_,
            jinja2.ext.do,
        ]
E       AttributeError: module 'jinja2.ext' has no attribute 'with_'

compressor/tests/test_offline.py:251: AttributeError
```

This is most likely due to an unsupported jinja2 version. Is it possible to update the jinja2 support or at least pin the version dependency?

Best regards
Jonas

**Comment by onny** [comment]
Possibly related https://github.com/django-compressor/django-compressor/pull/1125

**Comment by diox** [comment]
Yes, compressor is not compatible with the latest Jinja2. It's not an actual requirement - you can use it without the jinja extension, so we can't pin it (except in tests, [where we do](https://github.com/django-compressor/django-compressor/blob/d0b3d998c32eaa960f84a6ea41676ca338542777/requirements/tests.txt#L1)).

I'll gladly review patches fixing compatibility with Jinja 3.1.x (while keeping 3.0.x support).

**Comment by dvzrv** [comment]
Tests pass with jinja2 >= 3.1.0 when removing that line: https://github.com/django-compressor/django-compressor/blob/99a4145bbf41b0d1192319a84b83c9a609621cf8/compressor/tests/test_offline.py#L251

Seems `WithExtension` is now built-in: https://jinja.palletsprojects.com/en/3.1.x/changes/#version-3-1-0

@diox What is the basis of your assertion that django-compressor is not compatible with jinja2 >= 3.1.0? If it is only this failing test, it could be fixed by bumping jinja2. You seem to only make use of jinja2.ext.Extension otherwise and that is still available.

**Comment by diox** [comment]
You're right, I had not looked at this carefully enough. It does look like we can just remove that line and bump the version in test requirements.

**Comment by adamzap** [comment]
Tests are passing on `master` and locally for me, for what it's worth. Can this issue be closed?

https://github.com/django-compressor/django-compressor/blob/3d09a4cda1b9762005fc2e43cec7e185f9ee4510/requirements/tests.txt#L1

**Comment by diox** [comment]
I pushed a fix to `develop` but haven't made a release with it yet. Keeping this issue opened until I do.

**Comment by adamzap** [comment]
Ok, fair enough, thanks!

**Comment by dvzrv** [comment]
I guess this can be closed now? No modification is needed for e.g. 4.4.

**Comment by diox** [comment]
Yes, that was fixed in bcdd21956a841001b17a6a3d3b996bcbd696dbb4 which was released with 4.2.
